File: src/co/nums/intellij/aem/htl/file/HtlFileViewProvider.java
package co.nums.intellij.aem.htl.file;

import co.nums.intellij.aem.htl.HtlLanguage;
import co.nums.intellij.aem.htl.psi.HtlTokenTypes;
import com.intellij.lang.Language;
import com.intellij.lang.LanguageParserDefinitions;
import com.intellij.lang.ParserDefinition;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.LanguageSubstitutors;
import com.intellij.psi.MultiplePsiFilesPerDocumentFileViewProvider;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.impl.source.PsiFileImpl;
import com.intellij.psi.templateLanguages.ConfigurableTemplateLanguageFileViewProvider;
import com.intellij.psi.templateLanguages.TemplateDataElementType;
import com.intellij.psi.templateLanguages.TemplateDataLanguageMappings;
import com.intellij.util.containers.ContainerUtil;
import gnu.trove.THashSet;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Set;
import java.util.concurrent.ConcurrentMap;

public class HtlFileViewProvider extends MultiplePsiFilesPerDocumentFileViewProvider
		implements ConfigurableTemplateLanguageFileViewProvider {

	private static final ConcurrentMap<String, TemplateDataElementType> TEMPLATE_DATA_TO_LANG = ContainerUtil.newConcurrentMap();

	private final Language baseLanguage;
	private final Language templateLanguage;

	public HtlFileViewProvider(PsiManager manager, VirtualFile file, boolean physical, Language baseLanguage,
			Language templateLanguage) {
		super(manager, file, physical);
		this.baseLanguage = baseLanguage;
		this.templateLanguage = templateLanguage;
	}

	public HtlFileViewProvider(PsiManager manager, VirtualFile file, boolean physical, Language baseLanguage) {
		this(manager, file, physical, baseLanguage, getTemplateDataLanguage(manager, file));
	}

	@NotNull
	private static Language getTemplateDataLanguage(PsiManager manager, VirtualFile file) {
		Language dataLanguage = TemplateDataLanguageMappings.getInstance(manager.getProject()).getMapping(file);
		if (dataLanguage == null) {
			dataLanguage = HtlLanguage.getDefaultTemplateLang().getLanguage();
		}

		Language substituteLanguage = LanguageSubstitutors.INSTANCE.substituteLanguage(dataLanguage, file,
				manager.getProject());

		// only use a substituted language if it's templateable
		if (TemplateDataLanguageMappings.getTemplateableLanguages().contains(substituteLanguage)) {
			dataLanguage = substituteLanguage;
		}
		return dataLanguage;
	}

	@Override
	public boolean supportsIncrementalReparse(@NotNull Language rootLanguage) {
		return false;
	}

	@NotNull
	@Override
	public Language getBaseLanguage() {
		return baseLanguage;
	}

	@NotNull
	@Override
	public Language getTemplateDataLanguage() {
		return templateLanguage;
	}

	@NotNull
	@Override
	public Set<Language> getLanguages() {
		return new THashSet<>(Arrays.asList(new Language[] { baseLanguage, getTemplateDataLanguage() }));
	}

	@Override
	protected MultiplePsiFilesPerDocumentFileViewProvider cloneInner(VirtualFile virtualFile) {
		return new HtlFileViewProvider(getManager(), virtualFile, false, baseLanguage, templateLanguage);
	}

	@Override
	protected PsiFile createFile(@NotNull Language language) {
		ParserDefinition parserDefinition = getDefinition(language);
		if (parserDefinition == null) {
			return null;
		}

		if (language.is(getTemplateDataLanguage())) {
			PsiFileImpl file = (PsiFileImpl) parserDefinition.createFile(this);
			file.setContentElementType(getTemplateDataElementType(getBaseLanguage()));
			return file;
		} else if (language.isKindOf(getBaseLanguage())) {
			return parserDefinition.createFile(this);
		} else {
			return null;
		}
	}

	private ParserDefinition getDefinition(Language language) {
		ParserDefinition parserDefinition;
		if (language.isKindOf(getBaseLanguage())) {
			parserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(
					language.is(getBaseLanguage()) ? language : getBaseLanguage());
		} else {
			parserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(language);
		}
		return parserDefinition;
	}

	private static TemplateDataElementType getTemplateDataElementType(Language lang) {
		TemplateDataElementType result = TEMPLATE_DATA_TO_LANG.get(lang.getID());
		if (result != null) {
			return result;
		}
		TemplateDataElementType created = new TemplateDataElementType("HTL_TEMPLATE_DATA", lang,
				HtlTokenTypes.HTML_FRAGMENT, HtlTokenTypes.HTL_FRAGMENT);
		TemplateDataElementType prevValue = TEMPLATE_DATA_TO_LANG.putIfAbsent(lang.getID(), created);

		return prevValue == null ? created : prevValue;
	}

}

File: src/main/kotlin/co/nums/intellij/aem/htl/file/HtlFileViewProvider.kt
package co.nums.intellij.aem.htl.file

import co.nums.intellij.aem.htl.HtlLanguage
import co.nums.intellij.aem.htl.psi.*
import com.intellij.lang.*
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.*
import com.intellij.psi.impl.source.PsiFileImpl
import com.intellij.psi.templateLanguages.*
import com.intellij.util.containers.ContainerUtil

class HtlFileViewProvider @JvmOverloads constructor(private val psiManager: PsiManager,
                                                    file: VirtualFile,
                                                    physical: Boolean,
                                                    private val baseLanguage: Language,
                                                    private val templateDataLanguage: Language = getTemplateDataLanguage(psiManager, file))
    : MultiplePsiFilesPerDocumentFileViewProvider(psiManager, file, physical), ConfigurableTemplateLanguageFileViewProvider {

    companion object {
        private val HTL_FRAGMENT = HtlElementType("HTL_FRAGMENT")
        private val TEMPLATE_DATA_TO_LANG = ContainerUtil.newConcurrentMap<String, TemplateDataElementType>()
    }

    override fun getBaseLanguage() = baseLanguage

    override fun getTemplateDataLanguage() = templateDataLanguage

    override fun getLanguages(): Set<Language> = setOf(baseLanguage, templateDataLanguage)

    override fun supportsIncrementalReparse(rootLanguage: Language) = false

    override fun cloneInner(virtualFile: VirtualFile) =
            HtlFileViewProvider(psiManager, virtualFile, false, baseLanguage, templateDataLanguage)

    override fun createFile(language: Language): PsiFile? {
        val parserDefinition = getParserDefinition(language)
        return when {
            language === templateDataLanguage -> {
                val file = parserDefinition.createFile(this) as PsiFileImpl
                file.contentElementType = getTemplateDataElementType(baseLanguage)
                file
            }
            language.isKindOf(baseLanguage) -> parserDefinition.createFile(this)
            else -> null
        }
    }

    private fun getParserDefinition(language: Language): ParserDefinition {
        val parserLanguage = when {
            language === baseLanguage -> language
            language.isKindOf(baseLanguage) -> baseLanguage
            else -> language
        }
        return LanguageParserDefinitions.INSTANCE.forLanguage(parserLanguage)
    }

    private fun getTemplateDataElementType(language: Language): TemplateDataElementType {
        val result = TEMPLATE_DATA_TO_LANG[language.id]
        if (result != null) {
            return result
        }
        val created = TemplateDataElementType("HTL_TEMPLATE_DATA", language, HtlTypes.OUTER_TEXT, HTL_FRAGMENT)
        val prevValue = TEMPLATE_DATA_TO_LANG.putIfAbsent(language.id, created)

        return prevValue ?: created
    }

}

private fun getTemplateDataLanguage(psiManager: PsiManager, file: VirtualFile): Language {
    var dataLanguage = TemplateDataLanguageMappings.getInstance(psiManager.project)?.getMapping(file)
            ?: HtlLanguage.getTemplateLanguageFileType().language

    val substituteLanguage = LanguageSubstitutors.INSTANCE.substituteLanguage(dataLanguage, file, psiManager.project)

    // only use a substituted language if it's templateable
    if (TemplateDataLanguageMappings.getTemplateableLanguages().contains(substituteLanguage)) {
        dataLanguage = substituteLanguage
    }
    return dataLanguage
}
